swift/StableDiffusion/pipeline/ControlNet.swift

```diff
@@ -0,0 +1,136 @@
+    
+    /// Load resources.
+    public func loadResources() throws {
+        for model in models {
+            try model.loadResources()
+        }
+    }
+
+    /// Unload the underlying model to free up memory
+    public func unloadResources() {
+        for model in models {
+            model.unloadResources()
+        }
+    }
+    
+    /// Pre-warm resources
+    public func prewarmResources() throws {
+        // Override default to pre-warm each model
+        for model in models {
+            try model.loadResources()
+            model.unloadResources()
+        }
+    }
+    
+    var inputImageDescriptions: [MLFeatureDescription] {
+        models.map { model in
+            try! model.perform {
+                $0.modelDescription.inputDescriptionsByName["controlnet_cond"]!
+            }
+        }
+    }
+    
+    /// The expected shape of the models image input
+    public var inputImageShapes: [[Int]] {
+        inputImageDescriptions.map { desc in
+            desc.multiArrayConstraint!.shape.map { $0.intValue }
+        }
+    }
+    
+    /// Calculate additional inputs for Unet to generate intended image following provided images
+    ///
+    /// - Parameters:
+    ///   - latents: Batch of latent samples in an array
+    ///   - timeStep: Current diffusion timestep
+    ///   - hiddenStates: Hidden state to condition on
+    ///   - images: Images for each ControlNet
+    /// - Returns: Array of predicted noise residuals
+    func execute(
+        latents: [MLShapedArray<Float32>],
+        timeStep: Int,
+        hiddenStates: MLShapedArray<Float32>,
+        images: [MLShapedArray<Float32>]
+    ) throws -> [[String: MLShapedArray<Float32>]] {
+        // Match time step batch dimension to the model / latent samples
+        let t = MLShapedArray(scalars: [Float(timeStep), Float(timeStep)], shape: [2])
+        
+        var currentOutputs: [[String: MLMultiArray]] = []
+        
+        for (modelIndex, model) in models.enumerated() {
+            let inputs = try latents.map { latent in
+                let dict: [String: Any] = [
+                    "sample": MLMultiArray(latent),
+                    "timestep": MLMultiArray(t),
+                    "encoder_hidden_states": MLMultiArray(hiddenStates),
+                    "controlnet_cond": MLMultiArray(images[modelIndex])
+                ]
+                return try MLDictionaryFeatureProvider(dictionary: dict)
+            }
+            
+            let batch = MLArrayBatchProvider(array: inputs)
+            
+            let results = try model.perform {
+                try $0.predictions(fromBatch: batch)
+            }
+            
+            for n in 0..<results.count {
+                let result = results.features(at: n)
+                if currentOutputs.count < results.count {
+                    let initOutput = result.featureNames.reduce(into: [String: MLMultiArray]()) { output, k in
+                        output[k] = MLMultiArray(
```

Comment:
This would be a lot faster if we could pre-allocate the output with the expected size.

Reply:
Is this suggesting that we should pre-allocate MLShapedArray with a specific shape in output dictionary? If we do this before allocating model results, would we create an MLShapedArray filled with zero values?

Reply:
Yes, create it with the right size and fill with zeros.